sycl/source/detail/scheduler/commands.cpp

```diff
@@ -2416,19 +2417,50 @@ static pi_result SetKernelParamsAndLaunch(
   }
   if (OutEventImpl != nullptr)
     OutEventImpl->setHostEnqueueTime();
-  pi_result Error =
-      [&](auto... Args) {
-        if (IsCooperative) {
-          return Plugin
-              ->call_nocheck<PiApiKind::piextEnqueueCooperativeKernelLaunch>(
-                  Args...);
-        }
-        return Plugin->call_nocheck<PiApiKind::piEnqueueKernelLaunch>(Args...);
-      }(Queue->getHandleRef(), Kernel, NDRDesc.Dims, &NDRDesc.GlobalOffset[0],
-        &NDRDesc.GlobalSize[0], LocalSize, RawEvents.size(),
+  if (KernelUsesClusterLaunch) {
+    std::vector<pi_launch_property> property_list;
+
+    pi_launch_property_value launch_property_value_cluster_range;
+    launch_property_value_cluster_range.cluster_dims[0] =
+        NDRDesc.ClusterDimensions[0];
+    launch_property_value_cluster_range.cluster_dims[1] =
+        NDRDesc.ClusterDimensions[1];
+    launch_property_value_cluster_range.cluster_dims[2] =
+        NDRDesc.ClusterDimensions[2];
+
+    property_list.push_back(
+        {pi_launch_property_id::PI_LAUNCH_PROPERTY_CLUSTER_DIMENSION,
+         launch_property_value_cluster_range});
+
+    if (IsCooperative) {
+      pi_launch_property_value launch_property_value_cooperative;
+      launch_property_value_cooperative.cooperative = 1;
+      property_list.push_back(
+          {pi_launch_property_id::PI_LAUNCH_PROPERTY_COOPERATIVE,
+           launch_property_value_cooperative});
+    }
+
+    return Plugin->call_nocheck<PiApiKind::piextEnqueueKernelLaunchCustom>(
+        Queue->getHandleRef(), Kernel, NDRDesc.Dims, &NDRDesc.GlobalSize[0],
+        LocalSize, property_list.size(), property_list.data(), RawEvents.size(),
         RawEvents.empty() ? nullptr : &RawEvents[0],
         OutEventImpl ? &OutEventImpl->getHandleRef() : nullptr);
-  return Error;
+  } else {
+    pi_result Error =
+        [&](auto... Args) {
+          if (IsCooperative) {
+            return Plugin
+                ->call_nocheck<PiApiKind::piextEnqueueCooperativeKernelLaunch>(
+                    Args...);
+          }
+          return Plugin->call_nocheck<PiApiKind::piEnqueueKernelLaunch>(
+              Args...);
+        }(Queue->getHandleRef(), Kernel, NDRDesc.Dims, &NDRDesc.GlobalOffset[0],
```

Comment:
Using a lambda IMO is less readable than just doing:
```
if (IsCooperative)
  return Plugin->call_nocheck<PiApiKind::piextEnqueueCooperativeKernelLaunch>(
                    blah,blah,blah);
return Plugin->call_nocheck<PiApiKind::piEnqueueKernelLaunch>(
              blah,blah,blah);

Reply:
This was an existing code block, unrelated to my changes.
It came up in the diff because in got encapsulated in an else block now

Reply:
I'm aware of that. I think it should be changed.

Reply:
So this is an immediately evaluated anonymous variadic lambda used to avoid a simple ternary expression?
I agree this is *way* over complicated and makes reading it a hostile experience. I'd be happy to see it changed in a fixup commit, but given our squash and merge policy this is a bit awkward, so maybe it can wait for another PR, Hugh?

Reply:
OK sure yep feel free to leave this out.

Reply:
Also, this doesn't need to be in the diff. Since the `if` statement ends in a `return`, you don't need the `else`, which would remove this unmodified code from diff

Reply:
> Also, this doesn't need to be in the diff. Since the if statement ends in a return, you don't need the else, which would remove this unmodified code from diff

Sorry had missed this comment. Just realized why I had done it, so the code previous to my changes was 
```cpp
    pi_result Error =
        [&](auto... Args) {
          if (IsCooperative) {
            return Plugin
                ->call_nocheck<PiApiKind::piextEnqueueCooperativeKernelLaunch>(
                    Args...);
          }
          return Plugin->call_nocheck<PiApiKind::piEnqueueKernelLaunch>(
              Args...);
        }(Queue->getHandleRef(), Kernel, NDRDesc.Dims, &NDRDesc.GlobalOffset[0],
          &NDRDesc.GlobalSize[0], LocalSize, RawEvents.size(),
          RawEvents.empty() ? nullptr : &RawEvents[0],
          OutEventImpl ? &OutEventImpl->getHandleRef() : nullptr);
    return Error;

```

So if the kernel in not `isCooperative`, it would go ahead with a normal launch even if it is a cluster launch. To avoid that, I would need to check that it is not a cluster launch, and then have another if block beneath this lambda handling cluster launch...
Hence I handled cluster launch first, and then pushed the current one in an else block.

Reply:
Never mind this, I misunderstood your suggestion, which I thought was about moving the IIFE above this if block,

Removed the `else` in [96c84caf](https://github.com/intel/llvm/pull/14113/commits/96c84cafda173e6b236ad44bcc591aa34eed8bf4)